### ISSUE TITLE
[xla:gpu] with_sharding_constraint triggers crash on GPU

### ISSUE THREAD

**Issue by mattjj** [openxla/xla#10543]
Originally filed by @dlwh as google/jax#20149

> ### Description
> ```
> 2024-03-08 23:54:07.022991: F external/xla/xla/hlo/ir/hlo_instruction.h:1830] Check failed: has_sharding()
> ```
> 
> My non-minimized version gives a C++ stack trace, but this one doesn't. 🤷
> 
> ```
> import jax
> import jax.numpy as jnp
> from jax.sharding import Mesh, NamedSharding as N, PartitionSpec as P
> from jax.lax import with_sharding_constraint as shard
> 
> mesh = Mesh(jax.devices(), ('dp'))
> 
> model = jnp.zeros((320, 32))
> model = jax.device_put(model, N(mesh, P(None, 'dp')))
> 
> toks = jnp.zeros((32, 128), dtype=jnp.int32)
> 
> def foo(m, toks):
>     embeds = jnp.zeros((32, 128, 32))
>     logits = jnp.einsum('bld,md->blm', embeds, m)
>     targets = jax.nn.one_hot(toks, 320, dtype=jnp.float32)
>     # this wsc is what triggers the crash
>     targets = shard(targets, N(mesh, P('dp', None)))
>     #logits = -logits
>     entr = jnp.einsum('bld,bld->bl', targets, logits)
>     return entr.mean()
> 
> grad_foo = jax.value_and_grad(foo)
> 
> def step(model, y):
>     loss, grad = grad_foo(model, y)
>     return loss, grad
> 
> jit_step = jax.jit(step)
> 
> print(jit_step.lower(model, toks).as_text())
> 
> if __name__ == '__main__':
>     for i in range(10):
>         loss, grad = jit_step(model, toks)
>         model -= grad
>         print(loss)
> ```
> 
> ### System info (python version, jaxlib version, accelerator, etc.)
> ```
>  jax:    0.4.26.dev20240308
> jaxlib: 0.4.26.dev20240308
> numpy:  1.24.3
> python: 3.11.5 (main, Sep 11 2023, 13:54:46) [GCC 11.2.0]
> jax.devices (8 total, 8 local): [cuda(id=0) cuda(id=1) ... cuda(id=6) cuda(id=7)]
> process_count: 1
> platform: uname_result(system='Linux', node='radium-12696', release='5.4.0-173-generic', version='#191-Ubuntu SMP Fri Feb 2 13:55:07 UTC 2024', machine='x86_64')
> 
> 
> $ nvidia-smi
> Fri Mar  8 23:56:18 2024
> +---------------------------------------------------------------------------------------+
> | NVIDIA-SMI 535.161.07             Driver Version: 535.161.07   CUDA Version: 12.2     |
> |-----------------------------------------+----------------------+----------------------+
> | GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
> | Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
> |                                         |                      |               MIG M. |
> |=========================================+======================+======================|
> |   0  NVIDIA A100-SXM4-80GB          On  | 00000000:1B:00.0 Off |                    0 |
> | N/A   35C    P0              70W / 500W |    425MiB / 81920MiB |      3%      Default |
> |                                         |                      |             Disabled |
> +-----------------------------------------+----------------------+----------------------+
> |   1  NVIDIA A100-SXM4-80GB          On  | 00000000:3D:00.0 Off |                    0 |
> | N/A   34C    P0              78W / 500W |    425MiB / 81920MiB |      0%      Default |
> |                                         |                      |             Disabled |
> +-----------------------------------------+----------------------+----------------------+
> |   2  NVIDIA A100-SXM4-80GB          On  | 00000000:4E:00.0 Off |                    0 |
> | N/A   32C    P0              66W / 500W |    425MiB / 81920MiB |      1%      Default |
> |                                         |                      |             Disabled |
> +-----------------------------------------+----------------------+----------------------+
> |   3  NVIDIA A100-SXM4-80GB          On  | 00000000:5F:00.0 Off |                    0 |
> | N/A   34C    P0              73W / 500W |    425MiB / 81920MiB |      1%      Default |
> |                                         |                      |             Disabled |
> +-----------------------------------------+----------------------+----------------------+
> |   4  NVIDIA A100-SXM4-80GB          On  | 00000000:9D:00.0 Off |                    0 |
> | N/A   34C    P0              66W / 500W |    425MiB / 81920MiB |      0%      Default |
> |                                         |                      |             Disabled |
> +-----------------------------------------+----------------------+----------------------+
> |   5  NVIDIA A100-SXM4-80GB          On  | 00000000:BD:00.0 Off |                    0 |
> | N/A   35C    P0              70W / 500W |    425MiB / 81920MiB |      1%      Default |
> |                                         |                      |             Disabled |
> +-----------------------------------------+----------------------+----------------------+
> |   6  NVIDIA A100-SXM4-80GB          On  | 00000000:CD:00.0 Off |                    0 |
> | N/A   34C    P0              71W / 500W |    425MiB / 81920MiB |      2%      Default |
> |                                         |                      |             Disabled |
> +-----------------------------------------+----------------------+----------------------+
> |   7  NVIDIA A100-SXM4-80GB          On  | 00000000:DD:00.0 Off |                    0 |
> | N/A   32C    P0              67W / 500W |    425MiB / 81920MiB |      0%      Default |
> |                                         |                      |             Disabled |
> +-----------------------------------------+----------------------+----------------------+
> 
> +---------------------------------------------------------------------------------------+
> | Processes:                                                                            |
> |  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
> |        ID   ID                                                             Usage      |
> |=======================================================================================|
> |    0   N/A  N/A   1606490      C   python                                      416MiB |
> |    1   N/A  N/A   1606490      C   python                                      416MiB |
> |    2   N/A  N/A   1606490      C   python                                      416MiB |
> |    3   N/A  N/A   1606490      C   python                                      416MiB |
> |    4   N/A  N/A   1606490      C   python                                      416MiB |
> |    5   N/A  N/A   1606490      C   python                                      416MiB |
> |    6   N/A  N/A   1606490      C   python                                      416MiB |
> |    7   N/A  N/A   1606490      C   python                                      416MiB |
> +---------------------------------------------------------------------------------------+
> ```

Meta question: When I cross-file issues like this, would you be able to work much faster with an HLO reproducer, rather than JAX code? Or is sharing a JAX reproducer like this okay?

**Comment by dlwh** [comment]
Sorry I should have followed up on this. We could trigger this with N>=2 GPUs, but not N=1.

The node in question had a lot of problems and once it was reimaged and we switched to using a docker container with CUDA+JAX it went away. I dunno. I am about to lose access to the hardware in question so I can't even promise that I can reproduce it anymore. (I don't really understand why that would affect this particular part of the compiler though.)

**Comment by mattjj** [comment]
No problem!

So, do you think we should close it, or keep it open?

**Comment by dlwh** [comment]
probably? chances are if it's a real problem it'll come up again and maybe someone will be able to better reproduce it

**Comment by mattjj** [comment]
Sounds good. Maybe someone will find this issue and ping/reopen it.

**Comment by dlwh** [comment]
I'm back to being able to reproduce this.

```
(levanter) dlwh@sphinx4:~/src/levanter$ python -c 'import jax; jax.print_environment_info()'
jax:    0.4.25
jaxlib: 0.4.25
numpy:  1.26.2
python: 3.10.13 (main, Sep 11 2023, 13:44:35) [GCC 11.2.0]
jax.devices (2 total, 2 local): [cuda(id=0) cuda(id=1)]
process_count: 1
platform: uname_result(system='Linux', node='sphinx4.stanford.edu', release='5.15.0-41-generic', version='#44~20.04.1-Ubuntu SMP Fri Jun 24 13:27:29 UTC 2022', machine='x86_64')


$ nvidia-smi
Mon Mar 25 11:49:18 2024
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.54.03              Driver Version: 535.54.03    CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  NVIDIA A100-SXM4-80GB          On  | 00000000:07:00.0 Off |                    0 |
| N/A   37C    P0              71W / 350W |    429MiB / 81920MiB |      2%      Default |
|                                         |                      |             Disabled |
+-----------------------------------------+----------------------+----------------------+
|   1  NVIDIA A100-SXM4-80GB          On  | 00000000:4D:00.0 Off |                    0 |
| N/A   38C    P0              71W / 350W |    429MiB / 81920MiB |      0%      Default |
|                                         |                      |             Disabled |
+-----------------------------------------+----------------------+----------------------+

+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
|    0   N/A  N/A    692051      C   python                                      416MiB |
|    1   N/A  N/A    692051      C   python                                      416MiB |
+---------------------------------------------------------------------------------------+
```

**Comment by dlwh** [comment]
(totally different machine from before)

**Comment by dlwh** [comment]
reproed it on yet another machine. this time on google cloud! I think it's a driver thing

**Comment by cheshire** [comment]
> 2024-03-08 23:54:07.022991: F external/xla/xla/hlo/ir/hlo_instruction.h:1830] Check failed: has_sharding()

It's a compile-time crash, which should not depend on GPU.

**Comment by bixia1** [comment]
I can run your colab with 8 A100 GPUs without any problem. See the screenshot with the HLO produced and the loss printed from the loop.

![image](https://github.com/user-attachments/assets/4a1fd55e-6f5b-4ac8-9af5-a7e5c52005d3)

**Comment by dlwh** [comment]
we figured out how to work around and I have no idea if it reproduces at all right now

**Comment by bixia1** [comment]
Can you describe your work around?

**Comment by dlwh** [comment]
We used the nvidia docker container and it worked there for whatever reason…

I think something was misconfigured, but that only thing that made sense
was the cuda version.
